tests/test_api_requests.py

```diff
@@ -322,6 +233,1467 @@ async def parse(self, response):
+            "zyte_api_automap",
+        ),
+    ],
+)
+def test_default_params_immutability(arg_key, meta_key, setting, meta):
+    """Make sure that the merging of Zyte API parameters from the *arg_key*
+    _get_api_params parameter with those from the *meta_key* request metadata
+    key does not affect the contents of the setting for later requests."""
+    request = Request(url="https://example.com")
+    request.meta[meta_key] = meta
+    default_params = copy(setting)
+    _get_api_params(
+        request,
+        **{
+            **GET_API_PARAMS_KWARGS,
+            arg_key: default_params,
+        },
     )
+    assert default_params == setting
+
+
+def _test_automap(global_kwargs, request_kwargs, meta, expected, warnings, caplog):
+    request = Request(url="https://example.com", **request_kwargs)
+    request.meta["zyte_api_automap"] = meta
+    with caplog.at_level("WARNING"):
+        api_params = _get_api_params(
+            request,
+            **{
+                **GET_API_PARAMS_KWARGS,
+                **global_kwargs,
+                "transparent_mode": True,
+            },
+        )
+    assert api_params == expected
+    if warnings:
+        for warning in warnings:
+            assert warning in caplog.text
+    else:
+        assert not caplog.records
+
+
+@pytest.mark.parametrize(
+    "meta,expected,warnings",
+    [
+        # If no other known main output is specified in meta, httpResponseBody
+        # is requested.
+        ({}, {"httpResponseBody": True, "httpResponseHeaders": True}, []),
+        (
+            {"unknownMainOutput": True},
+            {
+                "httpResponseBody": True,
+                "httpResponseHeaders": True,
+                "unknownMainOutput": True,
+            },
+            [],
+        ),
+        # If httpResponseBody is unnecessarily requested in meta, a warning is
+        # logged.
+        (
+            {"httpResponseBody": True},
+            {"httpResponseBody": True, "httpResponseHeaders": True},
+            ["do not need to set httpResponseBody to True"],
+        ),
+        # If other main outputs are specified in meta, httpRequestBody is not
+        # set.
+        (
+            {"browserHtml": True},
+            {"browserHtml": True, "httpResponseHeaders": True},
+            [],
+        ),
+        (
+            {"screenshot": True},
+            {"screenshot": True},
+            [],
+        ),
+        (
+            {"browserHtml": True, "screenshot": True},
+            {"browserHtml": True, "httpResponseHeaders": True, "screenshot": True},
+            [],
+        ),
+        # If no known main output is specified, and httpResponseBody is
+        # explicitly set to False, httpResponseBody is unset and no main output
+        # is added.
+        (
+            {"httpResponseBody": False},
+            {},
+            [],
+        ),
+        (
+            {"httpResponseBody": False, "unknownMainOutput": True},
```

Comment:
I may have missed it as you may have a succinct test somewhere but to make sure, are there tests using `None` unset Zyte API params like these?

Reply:
There is `test_default_params_merging` to test that a param from a default params setting can be unset with `None`. But if you mean to test that `param: None` and `param: False` behave the same, there is no test.

Should we allow `None` to have the same effect as `False`? (no opinion either way from me, and I am not sure that is what you meant here)

Reply:
I was thinking of a test that ensures `param: None` would behave the same way as if `param` wasn't set at all. Having such a test could help avoid potential regressions down the line.